src/Template/PhpTemplate.php

```diff
@@ -0,0 +1,232 @@
+            static::VAR_CONTEXT => $this->getContextFunction($context),
+            static::VAR_FUNCTION => $this->getFunctionsFunction($context),
+        ];
+    }
+
+    /**
+     * Retrieves a function which itself when invoked retrieves a value from a context.
+     *
+     * That function MUST take a stringable key, and an optional default.
+     * It MUST return the value if the key is found.
+     * It MUST return the default value if the key is not found.
+     * It MAY throw {@see Exception} if there is a problem retrieving the value.
+     *
+     * @param array|ArrayAccess|ContainerInterface $context The context to get the retrieval function for.
+     *
+     * @throws Exception If problem retrieving.
+     *
+     * @return callable The function that retrieves a context value.
+     * function (string $key, mixed $default): mixed
+     */
+    protected function getContextFunction($context): callable
+    {
+        return function (string $key, $default = null) use ($context) {
+            return $this->getContextValue($context, $key, $default);
+        };
+    }
+
+    /**
+     * Retrieves a function which itself invokes a template function by name.
+     *
+     * That function MUST take a string key, and a variable list of arguments.
+     * It MUST invoke an underlying function mapped by name.
+     * It MUST pass these arguments to the underlying function , as a variable list.
+     * It MUST return the result of the invocation of the underlying function.
+     * It MAY throw an {@see Exception} if there is a problem invoking the underlying function.
+     *
+     * @param array|ArrayAccess|ContainerInterface $context The context to get the functions function for.
+     *
+     * @throws Exception If problem retrieving.
+     *
+     * @return callable The function that invokes a template function.
+     * function (string $funcName, ...$args): mixed
+     */
+    protected function getFunctionsFunction($context): callable
+    {
+        return function (string $funcName, ...$args) use ($context) {
+            return $this->invokeFunction($funcName, $args);
+        };
+    }
+
+    /**
+     * Retrieves a value from a context.
+     *
+     * @param array|ArrayAccess|ContainerInterface $context The context to retrieve the value from.
+     * @param string|Stringable $key The key to retrieve the value for.
+     * @param mixed $default The value to retrieve if the key is not found.
+     * @throws Exception If problem retrieving value.
+     *
+     * @return mixed The value that corresponds to the specified key in the context, or the default value if not found.
+     */
+    protected function getContextValue($context, string $key, $default = null)
+    {
+        try {
+            $this->validateContext($context);
+        } catch (InvalidArgumentException $e) {
+            throw new InvalidArgumentException($this->__('Could not retrieve value for key "%1$s": invalid context', $key), 0, $e);
+        }
```

Comment:
If the context is not allowed to change during rendering, then I suggest validating the context only once, before rendering begins. It allows potential context errors to be reported prior to rendering, as well as being reported even in cases where the template does not use the context.

The latter could be especially beneficial for the developer, in cases where for example an invalid context is given to a template that does not use the context, but some third-party overrides the template, say through some module system :wink:, and the overriding template _does_ use the context. That would trigger this exception, but the original author would have never reproduced it.

Reply:
The reason I validate it here is that it is only this part of the code that actually depends on the context being a specific thing. So, I don't really see how this method could work without this validation: if it is not one of the supported types, this method must throw.

I do agree, however, that it could be a good idea to validate the context before rendering begins. I would need to also allow it to be `null`, in addition to the supported types, in that particular case.

Reply:
Sounds good to me.

Reply:
Fixed. Now validating.